feat(frontend): modularize header and footer components into routing-ui package

ericsen-tsai opened the pull request:
## Summary

This PR implements a comprehensive modularization of header and footer UI components by extracting them from the `frontend` package into a new `@kids-reporter/routing-ui` package. This change improves code reusability, maintainability, and follows monorepo best practices by creating a dedicated UI package that can be shared across different applications.

**IMPORTANT**: This PR is based on #775 

## Changes

### New Package Creation
- **Created `@kids-reporter/routing-ui` package** with complete build, lint, and type-check infrastructure
- **Implemented babel build system** for efficient bundling with CommonJS outputs

### Component Extraction and Refactoring
- **Extracted Header components** from `frontend` to `routing-ui`:
  - `Header` (main component with props interface)
  - `DesktopHeader` and `DesktopHeaderCompact` 
  - `MobileHeader`
  - `HeaderMenu` with `HeaderMenuItem` and `HeaderMenuItemGroup`
  - `HeaderPostTitleSetter` (renamed from `PostTitleSetter`)
  - `HeaderContext` and `HeaderProvider`
- **Extracted Footer component** with configurable props for social media, menu items, donate URL, and privacy policy
- **Extracted shared UI components**: `Button` and `Input` components
- **Extracted utility functions**: `cn` (className utility) and `generateSocialMediaConfig`
- **Extracted custom hooks**: `useIsAtTop`, `useMediaQuery`, `useScrollLevel`
- **Extracted icon components**: All header-related icons (Hamburger, Search, Clear, Login, Settings, Social Media icons)

### Frontend Package Updates
- **Updated `package.json`** to include `@kids-reporter/routing-ui` dependency
- **Modified layout files** to use new `Header` and `Footer` components from `routing-ui`
- **Updated TypeScript configuration** to include `routing-ui` path aliases
- **Added CSS imports** for `routing-ui` styles
- **Removed duplicate components** and utilities that were moved to `routing-ui`
- **Updated constants** to use `MenuItem` type from `routing-ui`

### Code Quality Improvements
- **Added `eslint-plugin-simple-import-sort`** for consistent import ordering across all packages
- **Created `.prettierignore` files** for all packages to exclude build outputs and generated files
- **Reordered imports** across 200+ files for consistency
- **Updated ESLint configurations** to include import sorting rules
- **Added Prettier configurations** with Tailwind CSS plugin support

### Build System Enhancements
- **Updated root `package.json`** to include `routing-ui` in workspaces
- **Added build scripts** for `routing-ui` package
- **Enhanced lint-staged configuration** to include Prettier formatting
- **Updated `yarn.lock`** with new dependencies and resolved versions

### File Structure Changes
- **Deleted 4 files** from `frontend` package (moved to `routing-ui`)
- **Renamed 1 file** (`postcss.config.js` → `postcss.config.mjs`)
- **Created 15+ new files** in `routing-ui` package
- **Modified 200+ files** across all packages for import reordering and configuration updates

## Additional Notes

### Import Ordering and Build Optimization
- **Import ordering enforcement**: Added \`eslint-plugin-simple-import-sort\` to prevent build tool confusion from duplicate imports like:
  \`\`\`typescript
  import { a } from \"b\"
  import { c } from \"b\"
  \`\`\`
  This ensures build tools can properly optimize and deduplicate imports without confusion.

### Tailwind CSS Class Ordering
- **Prettier Tailwind CSS plugin**: Added \`prettier-plugin-tailwindcss\` to automatically sort Tailwind CSS classes for better readability and consistency
- **Improved developer experience**: Class ordering makes it easier to read and maintain CSS classes in components

### Potential Bugs and Considerations
- **Import path changes**: All imports of header/footer components in `frontend` now reference `@kids-reporter/routing-ui`
- **CSS dependencies**: Ensure `routing-ui/styles.css` is properly imported in consuming applications
- **Type compatibility**: Verify that `MenuItem` and `SocialMediaHrefs` types are compatible with existing usage
- **Build order**: The `routing-ui` package must be built before `frontend` package in CI/CD pipelines
- **Peer dependencies**: Ensure `tailwindcss` is available in consuming applications for proper styling


### Performance Impact
- **Positive**: Reduced bundle size in `frontend` package by extracting shared components
- **Positive**: Better tree-shaking opportunities with modular package structure
- **Neutral**: No significant runtime performance impact expected

## Reference(s)

- [Notion Page](https://www.notion.so/twreporter/header-footer-modularization-2788dbdca93280e5ab85f8c9f0b86546)

ericsen-tsai committed:
chore: refactor ESLint and Prettier configurations across multiple packages

- Removed legacy ESLint and Prettier configuration files.
- Introduced new ESLint configuration files for various packages with updated rules and plugins.
- Updated Babel configuration for improved compatibility with modern JavaScript features.
- Updated TypeScript configurations to enhance type checking and module resolution.
- Adjusted package.json files to reflect new dependencies and versions.

chore(root): update eslint and prettier configurations

- Update `prettier.config.mjs` to include tailwind sorting rules.
- Updated `eslint.base.config.mjs` and `eslint.config.mjs` to include sorting rules.
- Added format checking in lint-staged.

ericsen-tsai committed:
feat(routing-ui): introduce routing UI package with header and footer components

- Added a new `@kids-reporter/routing-ui` package containing reusable header and footer components.
- Integrated social media configuration generation for dynamic footer links.
- Updated frontend package to utilize new routing UI components and adjusted dependencies accordingly.
- Removed legacy navigation and header components from the frontend package.

ericsen-tsai committed:
chore(cms): sort import order

ericsen-tsai commented on packages/frontend/Dockerfile:
```diff
@@ -1,15 +1,26 @@
 
-COPY . /build/
+# Copy source code
+COPY . /build/packages/frontend/
```
It has been removed https://github.com/kids-reporter/kids-reporter-monorepo/pull/775#discussion_r2412962944

ericsen-tsai commented on packages/image-resizer/Dockerfile:
```diff
@@ -1,16 +1,22 @@
 
 # Copy the rest of the application code to the working directory
-COPY . .
+COPY . /app/packages/image-resizer/
```
It has been removed https://github.com/kids-reporter/kids-reporter-monorepo/pull/775#discussion_r2412962944

ericsen-tsai commented on packages/routing-ui/src/header/shared-components.tsx:
```diff
@@ -259,29 +267,31 @@ export function ActionButtons({
             <Link
               href={item.href}
-              className="py-1 prose-p1 text-neutral-900 font-bold! hover:text-red-400 transition-colors h-6 flex items-center"
+              className="py-1 prose-p1 text-neutral-900 font-bold! h-6 flex items-center transition-colors hover:text-red-400"
```
The syntax was updated in V4 https://tailwindcss.com/docs/styling-with-utility-classes#using-the-important-modifier

ericsen-tsai committed:
refactor(routing-ui): update package configuration and add post-build script

- Removed the "type" field from package.json.
- Added a post-build script to append 'use client' directive to output files.
- Updated CSS variable names for better clarity.
- Adjusted input component styles for improved consistency and readability.

ericsen-tsai committed:
chore(routing-ui): migrate to Babel

- Migrated the build process to Babel.
- Replaced Link components with anchor tags.
- Adjusted Icon from JSX to function components.

ericsen-tsai commented:
@nickhsine 
I already migrated build process from tsup to babel, thanks!

ericsen-tsai committed:
chore(root): rebase formatting and align babel version

nickhsine commented on packages/api-gateway/src/gql-proxy-mini-app.ts:
```diff
@@ -1,10 +1,11 @@
 import express from 'express'
 import { createProxyMiddleware } from 'http-proxy-middleware'
 
+import consts from './constants.js'
```
question:
這是 simple-import-sort 排序後的結果嗎？

ericsen-tsai replied:
Yes, but I used the default settings. If you have any idea of the import order, I can adjust it, thanks!

nickhsine replied:
我以為是照字母的 order 排序，但仔細看，simple-import-sort 會將 relative path 移到另外的區塊去 sort。
我對於這個排序沒有問題，謝謝。

nickhsine commented on packages/frontend/src/app/(sticky-header)/layout.tsx:
```diff
@@ -1,10 +1,25 @@
+        socialMediaHrefs={SOCIAL_MEDIA_ITEMS.map((item) => item.href)}
+        searchPlaceholder={SEARCH_PLACEHOLDER}
+        subscribeUrl={SUBSCRIBE_URL}
+        donateUrl={DONATE_URL}
```
thought:
當沒有提供 Header props 時，Header component 可以使用 default values，如此我們可以維持
```
<Header />
```
的寫法。

ericsen-tsai replied:
Sure! https://github.com/kids-reporter/kids-reporter-monorepo/commit/9bd602d3c0c4352dea46767f2d535bda2e48e615

nickhsine commented on packages/frontend/src/app/layout.tsx:
```diff
@@ -32,7 +38,12 @@ export default async function RootLayout({
+              additionalMenuItems={ADDITIONAL_MENU_ITEMS}
+              donateUrl={DONATE_URL}
+              privacyPolicyUrl={PRIVACY_POLICY}
+            />
```
同 Header 的留言，不曉得可不可以有預設的 props？

ericsen-tsai replied:
as above, thanks!
https://github.com/kids-reporter/kids-reporter-monorepo/commit/9bd602d3c0c4352dea46767f2d535bda2e48e615

nickhsine commented:
thought:
目前 lint-staged 在檢查程式碼時，是用 root 的 eslint config；但在每個 subpkg 底下，也有自己的 eslint config。
不確定這兩個 eslint config 是否會有打架的情況？
例如 root eslint config 認為可以過的，在 subpkg 底下跑 lint:check 卻會報錯？

lint-staged 是否要用「修改到的程式碼所屬的 eslint config 來檢查」？

ericsen-tsai committed:
feat(routing-ui): add default values for header and footer

ericsen-tsai committed:
chore(root): update lint-staged configuration for ESLint and Prettier across multiple packages

- Refined lint-staged settings to target specific package directories for ESLint.

ericsen-tsai committed:
chore(routing-ui): update hamburger icon color to use css variable

ericsen-tsai commented:
> thought: 目前 lint-staged 在檢查程式碼時，是用 root 的 eslint config；但在每個 subpkg 底下，也有自己的 eslint config。 不確定這兩個 eslint config 是否會有打架的情況？ 例如 root eslint config 認為可以過的，在 subpkg 底下跑 lint:check 卻會報錯？
> 
> lint-staged 是否要用「修改到的程式碼所屬的 eslint config 來檢查」？

You're right.
I changed the lint-staged config to run lint based on the package eslint config

https://github.com/kids-reporter/kids-reporter-monorepo/commit/bf3fb5858af8555e5c79ae225f52dc567971faad

nickhsine reviewed (approved):
LGTM :saluting_face:
